### PR TITLE
Add support for controlled digital lending of eligible materials

### PR DESCRIPTION
**Pull request by mejackreed**
Handle the digital lending interactions with Symphony:

- charging and discharging material to our pseudopatron (ensuring we only loan what symphony says is available; and not to confuse digital checkouts with physical loans)
- placing CDL marked-up holds for patrons (to connect them with a particular checkout, and to form a digital waitlist), and managing the waitlist
- provide a JWT token to stacks to handle authorization without needing to check with Symphony every request 
- provide availability information for the viewer

### CODE REVIEW

File: app/services/symphony_client.rb [hunk 1]
```diff
@@ -93,6 +93,40 @@ def place_hold(**params)
...
+      JSON.parse(response.body)
+    rescue JSON::ParserError
+      nil
+    end
+  end
+
+  def check_out_item(item_barcode, patron_barcode)
+    response = authenticated_request('/circulation/circRecord/checkOut', headers: headers, method: :post, json: {
+                                       itemBarcode: item_barcode,
+                                       patronBarcode: patron_barcode
```
[the remark this answers is not in the document]
Comment (shelleydoljack) [reply app/services/symphony_client.rb:121]: Yes, we only need to pass itemBarcode and patronBarcode to checkout an item. We need to pass an override for patrons with blocks (delinquent).

[the remark this answers is not in the document]
Comment (mejackreed) [reply app/services/symphony_client.rb:121]: @shelleydoljack can we just always send the override? Or does it need to only happen on that delinquent patron conditions?

[the remark this answers is not in the document]
Comment (dlrueda) [reply app/services/symphony_client.rb:121]: Chatted quickly to Sarah about this.  Yes, we (obvs?) will “check out” these items to blocked users.  Sarah says yes, we can just always send the override.

[the remark this answers is not in the document]
Comment (shelleydoljack) [reply app/services/symphony_client.rb:121]: We can always send in the CKOBLOCKS acknowledgement (I think, no matter if DELINQUENT or OKAY) but there is likely a different thing we need to send for when the patron is BLOCKED. I'll need to test this and get back to you.

[the remark this answers is not in the document]
Comment (mejackreed) [reply app/services/symphony_client.rb:123]: Previously we had discussed doing the CDL checkouts to pseudo patrons. If we do that, do we need to accommodate all of these cases still?

File: app/controllers/cdl_controller.rb [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    messages = Array.wrap(checkout&.dig('messageList')).map { |message| message.dig('message') }
+
+    if messages.present?
+      render json: { error: messages.join(' ') }.to_json, status: :internal_server_error
+      return
+    end
+
+    payload = {
+      barcode: barcode,
+      id: checkout_params['id'],
```
Comment (cbeer) [inline app/controllers/cdl_controller.rb:51]:
```suggestion
      aud: checkout_params['id'],
```

Comment (mejackreed) [reply app/controllers/cdl_controller.rb:51]: I'm going to add this in the refactor

File: app/services/cdl_checkout.rb [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+  def place_hold
+    symphony_client.place_hold(
+      comment: "CDL checkout for #{druid}",
+      fill_by_date: DateTime.now + 1.year,
+      patron_barcode: user.library_id,
+      item: {
+        call: { key: callkey, resource: '/catalog/call' },
+        holdType: 'TITLE'
+      }
+      # key: 'CDL' # TODO: What should this be?
```
Comment (cbeer) [inline app/services/cdl_checkout.rb:49]: Just so we remember:

```suggestion
      key: 'SUL'
```

File: app/models/catalog_info.rb [hunk 4]
```diff
@@ -27,4 +31,30 @@ def home_location
   def fields
     (@response || {}).dig('fields') || {}
   end
+
+  def callkey
+    fields.dig('call', 'key')
+  end
+
+  def loan_period
+    (fields.dig('itemCategory3', 'dig')&.scan(/^CDL-(\d+)H$/)&.flatten&.first&.to_i || 2).hours
```
Comment (cbeer) [inline app/models/catalog_info.rb:40]:
```suggestion
    (fields.dig('itemCategory3', 'key')&.scan(/^CDL-(\d+)H$/)&.flatten&.first&.to_i || 2).hours
```

File: app/jobs/cdl_waitlist_job.rb [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+    active_hold_records, remaining_holds = cdl_hold_records.partition do |record|
+      record.circ_record_key == circ_record_key
+    end
+
+    if active_hold_records.any?
+      active_hold_record = active_hold_records.first
+      symphony_client.cancel_hold(active_hold_record.key)
+      if active_hold_record.next_up_cdl?
+        comment = active_hold_record.comment.gsub('NEXT_UP', 'EXPIRED')
```
Comment (cbeer) [inline app/jobs/cdl_waitlist_job.rb:24]: For active holds that we're expiring, we should gsub ACTIVE => COMPLETE

File: app/models/ability.rb [hunk 6]
```diff
@@ -72,6 +72,12 @@ def initialize(user, token = nil)
     end
 
     cannot :create, Scan unless user.super_admin? || current_user_in_scan_pilot_group?
+
+    if Settings.cdl.enabled
+      can :checkin, :cdl if user.webauth_user?
+      can :checkout, :cdl if user.webauth_user?
+      can :availability, :cdl if true == true
```
Comment (cbeer) [inline app/models/ability.rb:79]:
```suggestion
      can :availability, :cdl
```

🤷‍♂️

File: spec/services/cdl_checkout_spec_spec.rb [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    ]
+  end
+
+  before do
+    allow(user).to receive(:patron).and_return(Patron.new({}))
+    allow(SymphonyClient).to receive(:new).and_return(symphony_client)
+  end
+
+  describe '#process_checkout' do
+    # rubocop:disable RSpec/EmptyExampleGroup
```
Comment (jkeck) [inline spec/services/cdl_checkout_spec_spec.rb:29]: 😿

**Pull request merged**